### DISCUSSION TITLE
possible mypy regression introduced by 10.2.x

### DISCUSSION THREAD

**Discussion by ssbarnea** [Textualize/rich#1243] [Q&A] [answered]
Once rich 10.2.0 was released mypy [started to fail](https://github.com/ansible-community/ansible-lint/actions?query=event%3Aschedule) with an error like:
```
src/ansiblelint/__main__.py:204: error: "object" not callable
Found 1 error in 1 file (checked 67 source files)
```
The line is just a [console.print(...)](https://github.com/ansible-community/ansible-lint/blob/master/src/ansiblelint/__main__.py#L204) does not appear to do anything wrong.

I tried to replicate the issue with a single file but I failed, it seems to be related to the fact that `console` instance is a module variable, initialized at https://github.com/ansible-community/ansible-lint/blob/master/src/ansiblelint/color.py#L24

The code seems to work fine with 10.2, only mypy seems to become upset. I did not raise a bug because I am not sure that is a real bug in rich.

Note that ceiling rich `<10.2` does work for the moment.

**Comment by willmcgugan** [comment, marked as answer]
I'd put `reveal_type(console)` and `reveal_type(console.print)` to see what Mypy thinks those objects are.

**Reply by ssbarnea** [reply]
```
src/ansiblelint/__main__.py:204: note: Revealed type is 'rich.console.Console'
src/ansiblelint/__main__.py:205: note: Revealed type is 'def (*objects: Any, *, sep: builtins.str =, end: builtins.str =, style: Union[builtins.str, rich.style.Style, None] =, justify: Union[Literal['default'], Literal['left'], Literal['center'], Literal['right'], Literal['full'], None] =, overflow: Union[Literal['fold'], Literal['crop'], Literal['ellipsis'], Literal['ignore'], None] =, no_wrap: Union[builtins.bool, None] =, emoji: Union[builtins.bool, None] =, markup: Union[builtins.bool, None] =, highlight: Union[builtins.bool, None] =, width: Union[builtins.int, None] =, height: Union[builtins.int, None] =, crop: builtins.bool =, soft_wrap: Union[builtins.bool, None] =)'
```

**Reply by willmcgugan** [reply]
Perhaps the error is coming from `_rule_format_map[options.format](rules)`. Try putting that in a line of its own.

**Reply by ssbarnea** [reply]
It seems that the error was coming from my formatter and had nothing to do with rich. Not sure why it started to complain only when rich was updated. 

https://github.com/ansible-community/ansible-lint/pull/1553/files is fixing the issue. Apparently what causes mypy to fail here is that it no longer considers `str` as being a `Callable`. I added a function and avoided using str and its seems to no longer complain.

Thanks for the help! I did not know about reveal_type a very useful type debugging tool.
